File: packages/ispf-driver-api/src/main/java/com/ispf/driver/TelemetryQuality.java
package com.ispf.driver;

import java.util.Locale;

/**
 * Normalized telemetry quality (BL-82, ADR-0025).
 */
public final class TelemetryQuality {

    public enum Level {
        GOOD,
        UNCERTAIN,
        BAD;

        public boolean isPlottable() {
            return this != BAD;
        }

        public String apiName() {
            return name();
        }
    }

    private TelemetryQuality() {
    }

    public static Level parse(String raw) {
        if (raw == null || raw.isBlank()) {
            return Level.GOOD;
        }
        return switch (raw.trim().toUpperCase(Locale.ROOT)) {
            case "GOOD" -> Level.GOOD;
            case "UNCERTAIN", "UNCERT" -> Level.UNCERTAIN;
            case "BAD" -> Level.BAD;
            default -> Level.UNCERTAIN;
        };
    }

    public static boolean isPlottable(String raw) {
        return parse(raw).isPlottable();
    }

    public static String normalize(Level level) {
        return level != null ? level.apiName() : Level.GOOD.apiName();
    }
}

File: apps/web-console/src/utils/telemetryQuality.ts
/** Normalized telemetry quality (ADR-0025, BL-82). */

export type TelemetryQualityLevel = "GOOD" | "UNCERTAIN" | "BAD";

export function parseTelemetryQuality(raw: unknown): TelemetryQualityLevel {
  if (typeof raw !== "string" || raw.trim() === "") {
    return "GOOD";
  }
  const normalized = raw.trim().toUpperCase();
  if (normalized === "BAD") {
    return "BAD";
  }
  if (normalized === "UNCERTAIN" || normalized === "UNCERT") {
    return "UNCERTAIN";
  }
  return "GOOD";
}

/** Whether chart/trend widgets should plot this sample (BAD → gap). */
export function isPlottableTelemetryQuality(raw: unknown): boolean {
  return parseTelemetryQuality(raw) !== "BAD";
}

export function readRowQuality(row: Record<string, unknown> | undefined): TelemetryQualityLevel | undefined {
  if (!row || !("quality" in row)) {
    return undefined;
  }
  return parseTelemetryQuality(row.quality);
}
